### PR TITLE
Aircrack-ng parsing speed up, fix segfaults

### PR DESCRIPTION
**Pull request by ChrisLundquist**
Current master / 1.3 release segfaults
```
$ lldb aircrack-ng /srv/wifi/Kismet-20180709-18-38-41-1.pcapdump
(lldb) target create "aircrack-ng"
Current executable set to 'aircrack-ng' (x86_64).
(lldb) settings set -- target.run-args  "/srv/wifi/Kismet-20180709-18-38-41-1.pcapdump"
(lldb) run
Process 13675 launched: '/usr/local/bin/aircrack-ng' (x86_64)
Opening /srv/wifi/Kismet-20180709-18-38-41-1.pcapdump
Process 13675 stoppedse wait...
* thread #2, name = 'aircrack-ng', stop reason = signal SIGSEGV: invalid address (fault address: 0x689a010f1e00)
    frame #0: aircrack-ng`check_thread(arg=<unavailable>) at aircrack-ng.c:2324
   2321			while (ap_cur != NULL)
   2322			{
   2323				if (!memcmp(ap_cur->bssid, bssid, 6)) break;
-> 2324	
   2325				ap_prv = ap_cur;
   2326				ap_cur = ap_cur->next;
   2327			}
(lldb) ^Cclundquist@nuc:~/aircrack-ng$ lldb ./src/aircrack-ng /srv/wifi/Kismet-20180709-18-38-41-1.pcapdump
(lldb) target create "./src/aircrack-ng"
Current executable set to './src/aircrack-ng' (x86_64).
(lldb) settings set -- target.run-args  "/srv/wifi/Kismet-20180709-18-38-41-1.pcapdump"
(lldb) run
Process 13692 launched: './src/aircrack-ng' (x86_64)
Opening /srv/wifi/Kismet-20180709-18-38-41-1.pcapdump
Process 13692 stoppedse wait...
* thread #2, name = 'aircrack-ng', stop reason = signal SIGSEGV: invalid address (fault address: 0xb)
    frame #0: aircrack-ng`ac_aplist_free(ap_1st=0x00007ffff00993e0) at aircrack-ng.c:364
   361 			while (ap_cur->st_1st != NULL)
   362 			{
   363 				st_tmp = ap_cur->st_1st;
-> 364 				ap_cur->st_1st = ap_cur->st_1st->next;
   365 				free(st_tmp);
   366 				st_tmp = NULL;
   367 			}
```

if we use a different file with 1.3 (debug symbols):
```
 time ./src/aircrack-ng /srv/wifi/home-02.pcapdump
# 45 minutes later I give up
13716 clundqu+  20   0  101832  10424   3944 S 100.0  0.0  45:28.37 ./src/aircrack-ng /srv/wifi/home-02.pcapdump 
```

This branch:
```
$ time ./src/aircrack-ng /srv/wifi/home-02.pcapdump
Opening /srv/wifi/home-02.pcapdump
Read 19740955 packets.

Index number of target network ? ^C
Quitting aircrack-ng...
pthread_join: Success

real	0m4.268s
user	0m1.437s
sys	0m1.235s
```

~TODO memory cleanup.~ done

### CODE REVIEW

**Comment by aircrack-ng** [comment]
Would it be possible to split it up in 2 PR: first one to fix the issue and the second one, adding AVL?

**Comment by ChrisLundquist** [comment]
they are incompatible fixes. One is an issue traversing a linked list. The other replaces it.

**Comment by aircrack-ng** [comment]
Could you fix the issues reported by travis/appveyor?

**Comment by jbenden** [comment]
FYI: All of the buildbot's are crashing with the newer session support of aircrack; so please be sure to check locally that this works OK.

**Comment by ChrisLundquist** [comment]
- Replace the linked_list with an AVL tree. Previously we spent a lot of time traversing the list.
- Fixes segfaults related to traversing the linked list
- Fixes race condition in ac_aplist_free causing segfault.
- Sort the includes for aircrack-ng.c
- clang-format
- Fixes signed/unsigned compares

This branch can process 500MM packets across multiple files in a reasonable amount of time now.

**Comment by ChrisLundquist** [comment]
@jbenden this is a pretty big change so we should probably merge in any fixes we need for CI.

I have done some testing locally tracking down the segfaults, but there are a ton of use cases and code paths I haven't tried / don't know about.

I also was trying to make `./configure CFLAGS="-Wall -Wextra -pedantic"` the default cflags but couldn't quite figure out the autotools side. Though, it looks like clang starts to complain about some of our macros.

**Comment by ChrisLundquist** [comment]
Ah I see: `FAIL: test-aircrack-ng-0005.sh`

Yeah, I haven't figured out the magic command to actually run these tests locally.

`test-driver` doesn't seem to work if I set the 3 things it wants, or more likely I'm doing it wrong.
The env vars don't seem to be set if I run them directly, etc.

What's the best way to run tests? I had hoped `make test` would do it, so I'm at a loss short of hacking the files

**Comment by ChrisLundquist** [comment]
looks like `make check`

**Comment by jbenden** [comment]
`make check`

**Comment by aircrack-ng** [comment]
Check the file `test/test-suite.log` for details.

The command in this case would be:

```
./src/aircrack-ng test/wep_64_ptw.cap -l /dev/null | grep "KEY FOUND" | grep "1F:1F:1F:1F:1F"
echo $?
```

**Comment by ChrisLundquist** [comment]
For some reason the WEP state (`ptw_vague`) isn't being setup. I need some food then I'll keep digging.


```
Process 22586 launched: './src/aircrack-ng' (x86_64)
Opening test/wep_64_ptw.cap
Read 65282 packets.

   #  BSSID              ESSID                     Encryption

   1  00:12:BF:12:32:29  Appart                    WEP (30566 IVs)

Choosing first network as target.

Opening test/wep_64_ptw.cap
1 potential targetsease wait...

Attack will be restarted every 5000 captured ivs.
Starting PTW attack with 0 ivs.
Process 22586 stopped
* thread #1, name = 'aircrack-ng', stop reason = signal SIGSEGV: invalid address (fault address: 0x270298)
    frame #0: aircrack-ng`PTW_computeKey(state=0x0000000000000000, keybuf="", keylen=5, testlimit=1000, bf=0x000055555577f820, validchars=0x0000555555857fd0, attacks=0) at aircrack-ptw-lib.c:762
   759 			exit(-1);
   760 		}
   761 	
-> 762 		if (!(attacks & NO_KLEIN))
   763 		{
   764 			// Try the original klein attack first
   765 			for (i = 0; i < keylen; i++)
(lldb) bt
* thread #1, name = 'aircrack-ng', stop reason = signal SIGSEGV: invalid address (fault address: 0x270298)
  * frame #0: aircrack-ng`PTW_computeKey(state=0x0000000000000000, keybuf="", keylen=5, testlimit=1000, bf=0x000055555577f820, validchars=0x0000555555857fd0, attacks=0) at aircrack-ptw-lib.c:762
    frame #1: aircrack-ng`main at aircrack-ng.c:5566
    frame #2: aircrack-ng`main(argc=<unavailable>, argv=<unavailable>) at aircrack-ng.c:6796
    frame #3: libc.so.6`__libc_start_main(main=(aircrack-ng`main at aircrack-ng.c:5623), argc=4, argv=0x00007fffffffe438, init=<unavailable>, fini=<unavailable>, rtld_fini=<unavailable>, stack_end=0x00007fffffffe428) at libc-start.c:310
    frame #4: 0x000055555555c25a aircrack-ng`_start + 42
```

```
(lldb) frame select 1
frame #1: aircrack-ng`main at aircrack-ng.c:5566
   5563			else
   5564			{
   5565				/* try 1000 40bit keys first, to find the key "instantly" and you don't need to wait for 104bit to fail */
-> 5566				if (PTW_computeKey(ap_cur->ptw_vague,
   5567								   wep.key,
   5568								   5,
   5569								   1000,
```

**Comment by jbenden** [comment]
I'd check the `state` variable, as it's NULL:

```
* frame #0: aircrack-ng`PTW_computeKey(state=0x0000000000000000, keybuf="", keylen=5, testlimit=1000, bf=0x000055555577f820, validchars=0x0000555555857fd0, attacks=0) at aircrack-ptw-lib.c:762
```

**Comment by ChrisLundquist** [comment]
yes, that's what I was seeing too and was aiming at with:
> For some reason the WEP state (ptw_vague) isn't being setup.

I'm just not sure *why* it isn't being initialized, and what is *supposed* to initialize it.

**Comment by ChrisLundquist** [comment]
It is "fixed", but it isn't super great. (I'd love to understand the original flow better, and the intent)

I *believe* in the old world we destroyed and rebuilt our data structure between invocations of `check_thread` and `read_thread` and now we keep the data, so we had a dependency of it getting reinitialized. I'm not 100% sure though because I don't completely follow why we have `check_thread` and `read_thread` and why they have so much duplicate code.

My model of the code flow is roughly:
- parse options
- open each file from argv, parse each in a thread looking for WPA handshakes/WEP IVs
- (do the above again in check_thread or read_thread?)
- pick a target, spin up cracking threads as appropriate
- find key and exit or exhaust dictionaries

**Comment by ChrisLundquist** [comment]
Let me know if anything else needs to be done here

**Comment by jbenden** [comment]
I'm not fond of mixing functionality and code formatting changes, but the changes do seem OK,

I think we should wait for @Mister-X- approval on merging this, though...

**Comment by ChrisLundquist** [comment]
> I'm not fond of mixing functionality and code formatting changes, but the changes do seem OK,

Yeah, from what I gathered `clang-format` was The Way To Happiness^TM here. So, if I formatted this using `clang-format`, I may actually have *fewer* format conflicts since we're using an auto-formatter.



> I think we should wait for @Mister-X- approval on merging this, though...

yep yep, this is a pretty involved change so I'd appreciate all the eyes possible.

**Comment by aircrack-ng** [comment]
Agreed with @jbenden regarding formatting changes, please use our `clang-format`.

You pushed a not-so-great fix in a set of great changes. We'll document `check_thread` and `read_thread` in the next few days (to also avoid confusion in the future), so you can improve/fix the fix.

**Comment by ChrisLundquist** [comment]
Yeah, this should be using your `.clang-format` unless I missed something in the usage.

**Comment by ChrisLundquist** [comment]
> You pushed a not-so-great fix in a set of great changes.

Yeah, I *think* that "had" to happen based on my understanding on the previous dependency of relying on the other thread. I could easily be missing something though.

**Comment by jbenden** [comment]
Clang format would be ran like: `clang-format -i -style=file <filename>` from the root of the project.

@Mister-X- We could have Travis verify all changes meet the formatting style, if you like the idea.

**Comment by ChrisLundquist** [comment]
```
clundquist@nuc:~/aircrack-ng$ clang-format -i --style=file src/aircrack-ng.c
clundquist@nuc:~/aircrack-ng$ git diff
clundquist@nuc:~/aircrack-ng$ clang-format -i src/aircrack-ng.c 
clundquist@nuc:~/aircrack-ng$ git diff
clundquist@nuc:~/aircrack-ng$ 
```

Yep, this seems to be using your format.

```
clundquist@nuc:~/aircrack-ng$ strace clang-format -i src/aircrack-ng.c  2>&1 | grep ".clang-format"
execve("/usr/bin/clang-format", ["clang-format", "-i", "src/aircrack-ng.c"], 0x7ffcf9db2220 /* 24 vars */) = 0
stat("/home/clundquist/aircrack-ng/src/.clang-format", 0x7ffdc6697e70) = -1 ENOENT (No such file or directory)
stat("/home/clundquist/aircrack-ng/src/_clang-format", 0x7ffdc6697e70) = -1 ENOENT (No such file or directory)
stat("/home/clundquist/aircrack-ng/.clang-format", {st_mode=S_IFREG|0664, st_size=857, ...}) = 0
openat(AT_FDCWD, "/home/clundquist/aircrack-ng/.clang-format", O_RDONLY|O_CLOEXEC) = 3
```
It looks for `.clang-format` by default

**Comment by ChrisLundquist** [comment]
Turns out our label `exit_main` was dead code.

I've fixed all but one leak detected by valgrind:
```
==16037== HEAP SUMMARY:
==16037==     in use at exit: 32 bytes in 1 blocks
==16037==   total heap usage: 105,104 allocs, 105,103 frees, 25,172,402 bytes allocated
==16037== 
==16037== 32 bytes in 1 blocks are still reachable in loss record 1 of 1
==16037==    at 0x4C31B25: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==16037==    by 0x56F17E4: _dlerror_run (dlerror.c:140)
==16037==    by 0x56F1050: dlopen@@GLIBC_2.2.5 (dlopen.c:87)
==16037==    by 0x417F33: ac_crypto_engine_loader_load (crypto_engine_loader.c:356)
==16037==    by 0x403B70: load_aircrack_crypto_dso (aircrack-ng.c:5657)
==16037==    by 0x403B70: main (aircrack-ng.c:5725)
```

I don't quite see it.
```
# crypto_engine_loader.c:356
356     module = dlopen (module_filename, RTLD_LAZY);
```

It looks like this should get cleaned up on error or by `ac_crypto_engine_loader_unload`

In other news, `--enable-asan` runs clean on a "open the file and control+C" test which is a nice win over master!

**Comment by ChrisLundquist** [comment]
appears to be a valgrind false positive:
https://stackoverflow.com/questions/1542457/memory-leak-reported-by-valgrind-in-dlopen

**Comment by jbenden** [comment]
@ChrisLundquist I'm fixing the BSD configure problem, stay tuned...

**Comment by jbenden** [comment]
@ChrisLundquist I've resolved the BSD configure problem.

**Comment by ChrisLundquist** [comment]
Just checking in and making sure there weren't any notes I missed.

**Comment by aircrack-ng** [comment]
both threads take care of reading the files and are very similar.

- `check_thread` is only used when preparing the list of networks for the user to choose from. It won't be called if the BSSID is already set
- `read_thread` is used when actually loading the files for cracking

File: src/aircrack-ng.c [hunk 1]
```diff
@@ -1391,6 +1394,28 @@ static void read_thread(void *arg)
 					break;
 				}
 			}
+			append_ap(ap_cur);
+		}
+		else
+		{
```
Comment (aircrack-ng) [inline src/aircrack-ng.c]: The fix looks good but in 6 months, we won't remember why this code is here (that look like a duplicate of the code above), so could you add a comment below this line with a small explanation of the issue and reference the URL where the issue is discovered (of the exact post in this PR)?

**Comment by ChrisLundquist** [comment]
I've simplified the area where we ensure `ap_cur` is setup for WEP.

**Comment by aircrack-ng** [comment]
Travis failed on a test. Are you sure it's a good idea to append before initializing PTW?

**Comment by ChrisLundquist** [comment]
> Are you sure it's a good idea to append before initializing PTW?

Yeah, that's why I had it the other way before. To be more atomic.

I can revert that last change and just leave a comment. Darn.

**Pull request merged**